README.md
# MISiS ACM VS Code helper utility

## Setup
1. Download the latest release on the [Releases](https://github.com/gbougakov/misis-helper/releases) page
2. Open the `.vsix` file to add the extension to VS Code
3. Run the `МИСиС ACM: Войти` command in the quick actions menu (`F1`)
4. Log in

## Features
### View problem statements
Run the `МИСиС ACM: Открыть задачу` command and select the problem to open it's statement.

### Submit solutions
Open the file with your solution, run the `МИСиС ACM: Отправить решение` command and select the problem and solution language to submit the solution.

Language autodetect coming soon.

&copy; [George Bougakov](https://gbougakov.dev) 2019

CHANGELOG.md
# Change Log

## v1.0.2

- Add ability to paste example tests into terminal

## v1.0.0

- Initial release


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
import fetch from "node-fetch";
import OpenStatementCommand from "./commands/openProblemStatement";
import LoginCommand from "./commands/login";
import SubmitCommand from "./commands/submitSolution";

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log("Congratulations, your extension \"misis-helper\" is now active!");

  // The command has been defined in the package.json file
  // Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	const openStatementCmd = new OpenStatementCommand(context);

  let disposable = vscode.commands.registerCommand(
    "dev.gbougakov.misis.openProblemStatement",
    openStatementCmd.run
  );

	context.subscriptions.push(disposable);

	const loginCmd = new LoginCommand(context);
	
	let disposable2 = vscode.commands.registerCommand(
    "dev.gbougakov.misis.login",
    loginCmd.run
  );

	context.subscriptions.push(disposable2);

	const submitCmd = new SubmitCommand(context);
	
	let disposable3 = vscode.commands.registerCommand(
    "dev.gbougakov.misis.sendSolution",
    submitCmd.run
  );

  context.subscriptions.push(disposable3);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/commands/login.ts
import * as vscode from "vscode";
import fetch from "node-fetch";
import getToken from "../utils/getToken";

class LoginCommand {
  context: vscode.ExtensionContext;
  constructor(context: vscode.ExtensionContext) {
    this.context = context;
    this.run = this.run.bind(this);
  }
  async run() {
    const login = await vscode.window.showInputBox({
      placeHolder: "Логин"
    });
    const password = await vscode.window.showInputBox({
      placeHolder: "Пароль",
      password: true
    });
    
    try {
      const token = await getToken(login || "nooneisstupidenoughtousethisasausername", password || "***");

      const {fullName} = await fetch("https://contest.misis.ru/api/user/me", {
        "headers": {
          "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0",
          "Cookie": `auth.token=${token}`
        }
      }).then(res => res.json());
      vscode.window.showInformationMessage(`Вы вошли как ${fullName}`);
      this.context.globalState.update("login", login);
      this.context.globalState.update("password", password);
    } catch(error) {
      vscode.window.showErrorMessage("Неправильный логин/пароль");
    }
    
  }
}

export default LoginCommand;

src/commands/submitSolution.ts
import * as vscode from "vscode";
import fetch from "node-fetch";
import getToken from "../utils/getToken";

class SubmitCommand {
  context: vscode.ExtensionContext;
  constructor(context: vscode.ExtensionContext) {
    this.context = context;
    this.run = this.run.bind(this);
  }
  async run() {
    if (
      !this.context.globalState.get("login") ||
      !this.context.globalState.get("password")
    ) {
      const action = await vscode.window.showErrorMessage(
        "Вы не вошли",
        "Войти"
      );
      if (!action) {
        return;
      }

      vscode.commands.executeCommand("dev.gbougakov.misis.login");

      return;
    }
    const token = await getToken(
      this.context.globalState.get("login") as string,
      this.context.globalState.get("password") as string
    );

    if (!vscode.window.activeTextEditor) {
      vscode.window.showErrorMessage("У вас не открыт файл");
      return;
    }

    const text = vscode.window.activeTextEditor!.document.getText();

    const { contests } = await fetch(
      "https://contest.misis.ru/api/contest/all?count=10&offset=0&category=all&sort=byStart&sort_order=desc&query=",
      {
        headers: {
          Cookie: "auth.token=" + token,
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
        },
        method: "GET"
      }
    ).then(res => res.json());

    const selectedContestName = await vscode.window.showQuickPick(
      contests.map((o: { name: any }) => o.name),
      {
        placeHolder: "Выберите контест"
      }
    );

    if (!selectedContestName) {
      return;
    }

    const selectedContest = contests.find(
      (o: { name: string }) => o.name === selectedContestName
    );

    const problems = await fetch(
      `https://contest.misis.ru/api/contest/${selectedContest.id}/problems`,
      {
        headers: {
          Cookie: "auth.token=" + token,
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
        },
        method: "GET"
      }
    ).then(res => res.json());

    const selectedProblem = await vscode.window.showQuickPick(
      problems.map((o: { title: string }) => o.title),
      {
        placeHolder: "Выберите задачу"
      }
    );

    if (!selectedProblem) {
      return;
    }

    const problem = problems.find(
      (o: { title: string }) => o.title === selectedProblem
    );

    const languages = await fetch(
      `https://contest.misis.ru/api/contest/${selectedContest.id}/problems/${problem.internalSymbolIndex}/languages`,
      {
        headers: {
          Cookie: "auth.token=" + token,
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
        },
        method: "GET"
      }
    ).then(res => res.json());

    const selectedLanguage = await vscode.window.showQuickPick(
      languages.map((o: { name: string }) => o.name),
      {
        placeHolder: "Выберите язык"
      }
    );

    if (!selectedLanguage) {
      return;
    }

    const language = languages.find(
      (o: { name: string }) => o.name === selectedLanguage
    );

    const solution = await fetch(
      `https://contest.misis.ru/api/contest/${selectedContest.id}/solutions`,
      {
        headers: {
          Cookie: "auth.token=" + token,
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0",
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          contestId: selectedContest.id,
          languageId: language.id,
          solution: text,
          symbolIndex: problem.internalSymbolIndex
        }),
        method: "POST"
      }
    ).then(res => res.json());
    let lastAnswer = undefined;
    let count = 0;
    while(!lastAnswer || (lastAnswer as {verdict: {name: string}}).verdict === null) {
      if (count > 500) {
        vscode.window.showErrorMessage("Вашему решению не был вынесен вердикт. Скорее всего, упал МИСиС ACM");
        return;
      }
      const result = await fetch(
        `https://contest.misis.ru/api/contest/${selectedContest.id}/solutions/${solution.id}/code`,
        {
          headers: {
            Cookie: "auth.token=" + token,
            "User-Agent":
              "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
          },
          method: "GET"
        }
      ).then(res => res.json());
      console.log(result);
      lastAnswer = result;
      count++;
    }
    if (lastAnswer.verdict.id === 1) {
      vscode.window.showInformationMessage("🎉 Решение прошло проверку");
    } else {
      vscode.window.showInformationMessage("😞 Решение не прошло проверку. Вердикт: " + lastAnswer.verdict.name);
    }
  }
}

export default SubmitCommand;


src/commands/openProblemStatement.ts
import fetch from "node-fetch";
import * as vscode from "vscode";
import getToken from "../utils/getToken";

class OpenStatementCommand {
  context: vscode.ExtensionContext;
  constructor(context: vscode.ExtensionContext) {
    this.context = context;
    this.run = this.run.bind(this);
  }

  async run() {
    if (!this.context.globalState.get("login") || !this.context.globalState.get("password")) {
      const action = await vscode.window.showErrorMessage("Вы не вошли", "Войти");
      if (!action) {
        return;
      }

      vscode.commands.executeCommand("dev.gbougakov.misis.login");

      return;
    }
    const token = await getToken(this.context.globalState.get("login") as string, this.context.globalState.get("password") as string);

    const {contests} = await fetch(
      "https://contest.misis.ru/api/contest/all?count=10&offset=0&category=all&sort=byStart&sort_order=desc&query=",
      {
        headers: {
          Cookie:
            "auth.token=" + token,
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
        },
        method: "GET"
      }
    ).then(res => res.json());
    
    const selectedContestName = await vscode.window.showQuickPick(contests.map((o: { name: any; }) => o.name), {
      placeHolder: "Выберите контест"
    });
    
    if (!selectedContestName) {
      return;
    }
    
    const selectedContest = contests.find((o: { name: string; }) => o.name === selectedContestName);
  
    const problems = await fetch(`https://contest.misis.ru/api/contest/${selectedContest.id}/problems`, {
      headers: {
        Cookie:
          "auth.token=" + token,
        "User-Agent":
          "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:71.0) Gecko/20100101 Firefox/71.0"
      },
      method: "GET"
    }).then(res => res.json());
  
    const selectedProblem = await vscode.window.showQuickPick(problems.map((o: { title: string; }) => o.title), {
      placeHolder: "Выберите задачу"
    });
    
    if (!selectedProblem) {
      return;
    }
  
    const problem = problems.find((o: { title: string; }) => o.title === selectedProblem);
  
    const panel = vscode.window.createWebviewPanel(
      "catCoding", // Identifies the type of the webview. Used internally
      problem.title, // Title of the panel displayed to the user
      vscode.ViewColumn.Two, // Editor column to show the new webview panel in.
      {
        enableScripts: true
      } // Webview options. More on these later.
    );

    panel.webview.onDidReceiveMessage(
      message => {
        switch (message.command) {
          case "terminal":
            if (!vscode.window.activeTerminal) {
              vscode.window.showErrorMessage("Сначала, запусти свой код в терминале");
              return;
            }
            
            vscode.window.activeTerminal!.sendText(message.text, true);
            return;
        }
      },
      undefined,
      this.context.subscriptions
    );
    
    panel.webview.html = `<style>
    .tbinfo {
      background: rgba(255, 255, 0, 0.2);
    }
    
    table, th, td {
      border: 1px solid black;
      padding: 5px;
    }
    </style>
    <script>
    window.onload = () => {
      const vscode = acquireVsCodeApi()

      const nodes = [].slice.call(document.querySelectorAll("td"))
      
      nodes.forEach(node => {
        console.log(node)
        node.onclick = e => {
          console.log()
          vscode.postMessage({
            command: 'terminal',
            text: e.target.innerHTML.replace('<br>', '\\n')
          })
        }
      })
      
      const tables = [].slice.call(document.querySelectorAll("table"))
      
      tables.forEach(table => {
        var para = document.createElement("td");
        var node = document.createTextNode("Нажмите на ячейку, чтобы вставить ее содержимое в терминал");
        para.appendChild(node);
        
        para.className = "tbinfo";
        para.colSpan = 3;
      
        table.insertBefore(para, table.childNodes[0])
      })
    }
    </script>` + problem.htmlStatement;
  }
}

export default OpenStatementCommand;